### ISSUE TITLE
multigrid_invert_test linked without -lcublas by default

### ISSUE THREAD

**Issue by kostrzewa** [lattice/quda#883]
I don't know if it's our installation here in Bonn, but when I try to compile the tests (latest develop) after this call of CMake:

```sh
MPI_HOME=/opt/openmpi-2.0.2a1-with-pmi
LD_LIBRARY_PATH=${MPI_HOME}/lib:${LD_LIBRARY_PATH} \
CXX=${MPI_HOME}/bin/mpicxx \
CC=${MPI_HOME}/bin/mpicc \
cmake \
-DCMAKE_CUDA_COMPILER=/usr/local/cuda-9.2/bin/nvcc \
-DCMAKE_INSTALL_PREFIX="/qbigwork2/bartek/libs/bleeding_edge/pascal/$(basename $(pwd))" \
-DMPI_CXX_COMPILER=${MPI_HOME}/bin/mpicxx \
-DMPI_C_COMPILER=${MPI_HOME}/bin/mpicc \
-DQUDA_BUILD_SHAREDLIB=ON \
-DQUDA_MAX_MULTI_BLAS_N=9 \
-DQUDA_BUILD_ALL_TESTS=YES \
-DQUDA_GPU_ARCH=sm_60 \
-DQUDA_INTERFACE_QDP=ON \
-DQUDA_INTERFACE_MILC=OFF \
-DQUDA_MPI=OFF \
-DQUDA_DIRAC_WILSON=ON \
-DQUDA_DIRAC_TWISTED_MASS=ON \
-DQUDA_DIRAC_TWISTED_CLOVER=ON \
-DQUDA_DIRAC_NDEG_TWISTED_MASS=ON \
-DQUDA_DIRAC_CLOVER=ON \
-DQUDA_DYNAMIC_CLOVER=ON \
-DQUDA_DIRAC_DOMAIN_WALL=OFF \
-DQUDA_MULTIGRID=ON \
-DQUDA_TEX=OFF \
-DQUDA_USE_EIGEN=ON \
-DQUDA_DOWNLOAD_EIGEN=ON \
-DQUDA_DOWNLOAD_USQCD=ON \
-DQUDA_QIO=ON \
-DQUDA_QMP=ON \
-DQUDA_OPENMP=ON \
-DQUDA_BLOCKSOLVER=ON \
-DQUDA_DIRAC_STAGGERED=OFF /qbigwork2/bartek/code/bleeding_edge/quda_develop
```

Linking of ```multigrid_invert_test``` fails due to ```-lcublas``` not being passed as a linker flag (This is CUDA-9.2 installed in /usr/local/cuda-9.2 with all symlinks created).

I added ```-lcublas``` by setting ```-DCMAKE_EXE_LINKER_FLAGS``` as a workaround, but is that how it's supposed to behave?

**Comment by mathiaswagner** [comment]
No, obviously not what should happen. Do you mind running the cmake command in a fresh dir (without the `CMAKE_EXE_LINKER_FLAGS`) and send the CMakeCache?

Not sure yet where this goes wrong. I think the cache should include the cmake version but please also send that.

Thanks.

**Comment by kostrzewa** [comment]
[CMakeCache_quda_gh883_v01.txt](https://github.com/lattice/quda/files/3541622/CMakeCache_quda_gh883_v01.txt)

**Comment by kostrzewa** [comment]
```cmake version 3.13.2```

**Comment by kostrzewa** [comment]
I've reproduced this behaviour on PIzDaint, btw.

**Comment by kostrzewa** [comment]
```
[ 97%] Linking CUDA executable invert_test
[ 98%] Linking CUDA executable multigrid_invert_test
[ 98%] Linking CUDA device code CMakeFiles/dslash_ctest.dir/cmake_device_link.o
[ 98%] Linking CUDA executable blas_test
[ 98%] Linking CUDA executable dslash_ctest
/usr/bin/ld: ../lib/libquda.a(blas_cublas.cu.o): in function `quda::cublas::init()':
/users/bartek/code/2019_08_27/quda_develop/lib/blas_cublas.cu:31: undefined reference to `cublasCreate_v2'
/usr/bin/ld: ../lib/libquda.a(blas_cublas.cu.o): in function `quda::cublas::destroy()':
/users/bartek/code/2019_08_27/quda_develop/lib/blas_cublas.cu:38: undefined reference to `cublasDestroy_v2'
/usr/bin/ld: ../lib/libquda.a(blas_cublas.cu.o): in function `quda::cublas::BatchInvertMatrix(void*, void*, int, int, QudaPrecision_s, QudaFieldLocation_s)':
/users/bartek/code/2019_08_27/quda_develop/lib/blas_cublas.cu:75: undefined reference to `cublasCgetrfBatched'
/usr/bin/ld: /users/bartek/code/2019_08_27/quda_develop/lib/blas_cublas.cu:90: undefined reference to `cublasCgetriBatched'
/usr/bin/ld: link errors found, deleting executable `su3_test'
/usr/bin/X11/sha1sum: su3_test: No such file or directory
collect2: error: ld returned 1 exit status
```

**Comment by maddyscientist** [comment]
@kostrzewa after running `cmake` can you try running `ccmake`, re-generate the `CMakeCache.txt` file using `ccmake` and then try to recompile again.  I don't know why this fixes the issue, but this problem has been seen by others, e.g., at Jülich.

If this does fix the problem, could you send us the generated CMakeCache.txt file before and after running `ccmake`?

@mathiaswagner Did you ever diagnose this issue?

**Comment by kostrzewa** [comment]
@maddyscientist Sure thing. First of all, my list of loaded modules:

```
 $ module list
Currently Loaded Modulefiles:
  1) modules/3.2.10.6                                 17) job/2.2.3-6.0.7.1_5.43__g6c4e934.ari
  2) cray-mpich/7.7.2                                 18) dvs/2.7_2.2.118-6.0.7.1_10.2__g58b37a2
  3) slurm/17.11.12.cscs-1                            19) alps/6.6.43-6.0.7.1_5.45__ga796da32.ari
  4) xalt/daint-2016.11                               20) rca/2.2.18-6.0.7.1_5.47__g2aa4f39.ari
  5) daint-gpu                                        21) atp/2.1.2
  6) gcc/6.2.0                                        22) perftools-base/7.0.3
  7) craype-haswell                                   23) PrgEnv-gnu/6.0.4
  8) craype-network-aries                             24) CMake/3.12.4
  9) craype/2.5.15                                    25) CrayGNU/.18.08
 10) cray-libsci/18.07.1                              26) bzip2/.1.0.6-CrayGNU-18.08
 11) udreg/2.3.2-6.0.7.1_5.13__g5196236.ari           27) zlib/.1.2.11-CrayGNU-18.08
 12) ugni/6.0.14.0-6.0.7.1_3.13__gea11d3d.ari         28) cray-python/3.6.5.1
 13) pmi/5.0.14                                       29) Boost/1.67.0-CrayGNU-18.08-python3
 14) dmapp/7.1.1-6.0.7.1_5.45__g5a674e0.ari           30) cudatoolkit/9.1.85_3.18-6.0.7.0_5.1__g2eb7c52
 15) gni-headers/5.0.12.0-6.0.7.1_3.11__g3b1768f.ari  31) cray-hdf5/1.10.2.0
 16) xpmem/2.2.15-6.0.7.1_5.11__g7549d06.ari
```

CMakeCache.txt before and after ccmake will follow shortly!

**Comment by kostrzewa** [comment]
Before ccmake:

[CMakeCache_quda_gh883_PizDaint_before_ccmake.txt](https://github.com/lattice/quda/files/3556167/CMakeCache_quda_gh883_PizDaint_before_ccmake.txt)

**Comment by kostrzewa** [comment]
sorry, mistakenly had the ```CMAKE_EXE_LINKER_FLAGS``` in the call. Just a second.

**Comment by maddyscientist** [comment]
ping...did `ccmake` fix the issue @kostrzewa?

**Comment by kostrzewa** [comment]
No, see the two versions of CMakeCache in comment above yours: https://github.com/lattice/quda/issues/883#issuecomment-526225515

**Comment by kostrzewa** [comment]
Oh, I see, I never followed up with a report on the "I'll compile now". When the machine is out of maintenance tomorrow I'll check again, I seem to remember that it didn't work, but can't guarantee that I remember right.

**Comment by kostrzewa** [comment]
Alright, so it's indeed true that running ```ccmake .``` allows the tests to be linked correctly.

While the CMakeCache.txt is unaffected as you can see above in https://github.com/lattice/quda/issues/883#issuecomment-526225515, what does change are the files in 

```
${builddir}/tests/CMakeFiles/${testname}.dir/build.make
${builddir}/tests/CMakeFiles/${testname}.dir/link.txt
```

which obtain lines:

```build.make:tests/CMakeFiles/multigrid_invert_test.dir/cmake_device_link.o: /opt/nvidia/cudatoolkit9.1/9.1.85_3.18-6.0.7.0_5.1__g2eb7c52/lib64/stubs/libcublas.so
```

```
build.make:tests/multigrid_invert_test: /opt/nvidia/cudatoolkit9.1/9.1.85_3.18-6.0.7.0_5.1__g2eb7c52/lib64/stubs/libcublas.so
```

and the link command:

```
link.txt:/opt/cray/pe/craype/2.5.15/bin/CC  CMakeFiles/multigrid_invert_test.dir/multigrid_invert_test.cpp.o CMakeFiles/multigrid_invert_test.dir/wilson_dslash_reference.cpp.o CMakeFiles/multigrid_invert_test.dir/clover_reference.cpp.o CMakeFiles/multigrid_invert_test.dir/domain_wall_dslash_reference.cpp.o CMakeFiles/multigrid_invert_test.dir/blas_reference.cpp.o CMakeFiles/multigrid_invert_test.dir/cmake_device_link.o -o multigrid_invert_test -lcudadevrt -lcudart_static -lrt -lpthread -ldl -lcupti -lcudart -lcuda -lAtpSigHandler -lAtpSigHCommData -lrca -lhdf5_hl -lhdf5 -lhdf5_hl_cpp -lhdf5_cpp -lsci_gnu_61_mpi -lsci_gnu_61 -lmpich_gnu_51 -lmpichcxx_gnu_51 -lgfortran -lquadmath -lmvec -lm -lpthread -lstdc++ -lm -lgcc_s -lgcc -lc -lgcc_s -lgcc ../lib/libquda.a libquda_test.a /opt/nvidia/cudatoolkit9.1/9.1.85_3.18-6.0.7.0_5.1__g2eb7c52/lib64/stubs/libcublas.so -L/users/bartek/build/2019_08_27/quda_develop-dynamic_clover-553c60ccaff0211923ddcf896869098cc6229a0d_gh883_test4/usqcd/lib -lqio -llime -lqmp /opt/nvidia/cudatoolkit9.1/9.1.85_3.18-6.0.7.0_5.1__g2eb7c52/lib64/stubs/libcuda.so -lcudadevrt -lcudart_static -lrt -lpthread -ldl -lcupti -lcudart -lcuda -lAtpSigHandler -lAtpSigHCommData -lrca -lhdf5_hl -lhdf5 -lhdf5_hl_cpp -lhdf5_cpp -lsci_gnu_61_mpi -lsci_gnu_61 -lmpich_gnu_51 -lmpichcxx_gnu_51 -lgfortran -lquadmath -lmvec -lm -lstdc++ -lpthread -ldl -lcupti -lcudart -lcuda -lAtpSigHandler -lAtpSigHCommData -lrca -lhdf5_hl -lhdf5 -lhdf5_hl_cpp -lhdf5_cpp -lsci_gnu_61_mpi -lsci_gnu_61 -lmpich_gnu_51 -lmpichcxx_gnu_51 -lgfortran -lquadmath -lmvec -lm -lstdc++ -lgcc_s -lgcc -lc -lgcc_s -lgcc -lc  -L"/opt/nvidia/cudatoolkit9.1/9.1.85_3.18-6.0.7.0_5.1__g2eb7c52/lib64/stubs" -L"/opt/nvidia/cudatoolkit9.1/9.1.85_3.18-6.0.7.0_5.1__g2eb7c52/lib64" -lcudadevrt -lcudart_static -lrt -lpthread -ldl
```

where libcublas is explicitly listed.

**Comment by mathiaswagner** [comment]
Still no root cause but just wanted to add here that running `cmake .` after the initial cmake run also seems to work this around (no need to use `ccmake`).

**Comment by maddyscientist** [comment]
@mathiaswagner can you up this as a known issue on the QUDA CMake wiki page and the known work arounds?  I think then we can close this issue since it’s a CMake issue and not a QUDA issue.

**Comment by mathiaswagner** [comment]
Maybe #913 fixes things

**Comment by kostrzewa** [comment]
Yes, #913 works

```
/opt/openmpi-2.0.2a1-with-pmi/bin/mpicxx  CMakeFiles/multigrid_invert_test.dir/multigrid_invert_test.cpp.o CMakeFiles/multigrid_invert_test.dir/wilson_dslash_reference.cpp.o CMakeFiles/multigrid_invert_test.dir/clover_reference.cpp.o CMakeFiles/multigrid_invert_test.dir/domain_wall_dslash_reference.cpp.o CMakeFiles/multigrid_invert_test.dir/blas_reference.cpp.o -o multigrid_invert_test -Wl,-rpath,/hadron/bartek/build/bleeding_edge/pascal/quda_cmake_qublas/lib -lcudadevrt -lcudart_static -lrt -lpthread -ldl -lmpi_cxx -lmpi -lstdc++ -lm -lgcc_s -lgcc -lpthread -lc -lgcc_s -lgcc ../lib/libquda.so libquda_test.a /usr/local/cuda-9.2/lib64/stubs/libcublas.so -L/hadron/bartek/build/bleeding_edge/pascal/quda_cmake_qublas/usqcd/lib -lqio -llime -lqmp /usr/local/cuda-9.2/lib64/stubs/libcuda.so -lcudadevrt -lcudart_static -lrt -lpthread -ldl -lmpi_cxx -lmpi -lstdc++ -lm -lgcc_s -lgcc -lc -lpthread -ldl -lmpi_cxx -lmpi -lstdc++ -lm -lgcc_s -lgcc -lc  -L"/usr/local/cuda-9.2/lib64/stubs" -L"/usr/local/cuda-9.2/lib64" -lcudadevrt -lcudart_static -lrt -lpthread -ldl
```

**Comment by mathiaswagner** [comment]
Will be closed in develop once #919 is merged.
